File: android/src/main/kotlin/com/superwall/superwallkit_flutter/bridges/PaywallPresentationHandlerProxyBridge.kt
package com.superwall.superwallkit_flutter.bridges

import android.content.Context
import com.superwall.sdk.paywall.presentation.PaywallPresentationHandler
import com.superwall.superwallkit_flutter.invokeMethodOnMain
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import toJson

class PaywallPresentationHandlerProxyBridge(
    context: Context,
    bridgeId: BridgeId,
    initializationArgs: Map<String, Any>? = null,
) : BridgeInstance(context, bridgeId, initializationArgs) {
    val scope = CoroutineScope(Dispatchers.IO)

    companion object {
        fun bridgeClass(): BridgeClass = "PaywallPresentationHandlerProxyBridge"
    }

    val handler: PaywallPresentationHandler by lazy {
        PaywallPresentationHandler().apply {
            onPresent {
                scope.launch {
                    communicator().invokeMethodOnMain(
                        "onPresent",
                        mapOf("paywallInfoBridgeId" to it.createBridgeId())
                    )
                }
            }
            onDismiss { info, result ->
                scope.launch {
                    communicator().invokeMethodOnMain(
                        "onDismiss",
                        mapOf("paywallInfoBridgeId" to info.createBridgeId(),
                            "paywallResult" to result.toJson())
                    )
                }
            }

            onError {
                scope.launch {
                    communicator().invokeMethodOnMain(
                        "onError",
                        mapOf("errorString" to it.toString())
                    )
                }
            }
            onSkip {
                scope.launch {
                    communicator().invokeMethodOnMain(
                        "onSkip",
                        mapOf("paywallSkippedReasonBridgeId" to it.createBridgeId())
                    )
                }
            }
        }
    }
}

File: ios/Classes/Bridges/PaywallPresentationHandlerProxyBridge.swift
import Flutter
import SuperwallKit

public class PaywallPresentationHandlerProxyBridge: BridgeInstance {
  override class var bridgeClass: BridgeClass { "PaywallPresentationHandlerProxyBridge" }

  lazy var handler: PaywallPresentationHandler = {
    let handler = PaywallPresentationHandler()

    handler.onPresent { [weak self] paywallInfo in
      guard let self else { return }
      communicator.invokeMethodOnMain("onPresent", arguments: ["paywallInfoBridgeId": paywallInfo.createBridgeId()])
    }

    handler.onDismiss { [weak self] paywallInfo in
      guard let self else { return }
      communicator.invokeMethodOnMain("onDismiss", arguments: ["paywallInfoBridgeId": paywallInfo.createBridgeId()])
    }

    handler.onError { [weak self] error in
      guard let self else { return }
      communicator.invokeMethodOnMain("onError", arguments: ["errorString": error.localizedDescription])
    }

    handler.onSkip { [weak self] paywallSkippedReason in
      guard let self else { return }
      communicator.invokeMethodOnMain("onSkip", arguments: ["paywallSkippedReasonBridgeId": paywallSkippedReason.createBridgeId()])
    }

    return handler
  }()
}
